README.md
# ritdu-slacker
The Ringier SA internal interface for slack


ChangeLog
CHANGES
=======

v1.2.0
------

* Cleanup

v1.1.0
------

* Refactor into loadable class (#7)

v1.0.0
------

* v1.0.0
* Add all features

v0.0.1
------

* v0.0.1
* Add requests

v0.0.0
------

* --
* Initial PR
* Initial commit


tests/conftest.py
import responses

from pyfakefs.fake_filesystem_unittest import Patcher
from pytest import fixture

from ritdu_slacker.cli import SlackMessageCLI


@fixture
def mocked_responses():
    with responses.RequestsMock() as rsps:
        yield rsps


@fixture
def cli():
    yield SlackMessageCLI()


@fixture
def filesystem():
    with Patcher(allow_root_user=False) as patcher:
        yield patcher.fs


tests/test_cli.py
import json
import sys

import pytest
import responses


def test_no_arguments(capsys, cli):
    sys.argv = [
        "ritdu-slacker",
    ]

    with pytest.raises(SystemExit):
        cli.main()

    out, _ = capsys.readouterr()

    assert (
        out
        == """Usage: ritdu-slacker [OPTIONS] COMMAND [ARGS]...

  CLI tool send/update slack messages and send files

Options:
  -h, --help  Show this message and exit.
  --version   Show the version and exit.

Commands:
  file     Command to send file to thread
  message  Command to send message, reply to thread or reply and...
"""
    )


def test_message_slack_channel_no_thread(capsys, cli, mocked_responses):
    sys.argv = [
        "ritdu-slacker",
        "message",
        "--workspace",
        "myworkspace",
        "--channel",
        "mychannel",
        "--text",
        "Update in progress",
        "--thread-uuid",
        "30D06A8F-B6BE-43BD-BE75-8716C26C1102",
        "--thread-broadcast",
    ]

    response_payload = {
        "ok": True,
        "channel": "C123456",
        "message": {
            "text": "Update in progress",
            "type": "message",
        },
    }

    mocked_responses.add(
        responses.POST,
        "https://slacker.cube-services.net/api/message-template",
        json=response_payload,
    )

    with pytest.raises(SystemExit):
        cli.main()

    out, _ = capsys.readouterr()

    assert out == json.dumps(response_payload) + "\n"


def test_message_replace_message(capsys, cli, mocked_responses):
    sys.argv = [
        "ritdu-slacker",
        "message",
        "--workspace",
        "myworkspace",
        "--channel",
        "mychannel",
        "--text",
        "Update in progress",
        "--message-uuid",
        "30D06A8F-B6BE-43BD-BE75-8716C26C1102",
    ]

    response_payload = {
        "ok": True,
        "channel": "C123456",
        "message": {
            "text": "Update in progress",
            "type": "message",
        },
    }

    mocked_responses.add(
        responses.POST,
        "https://slacker.cube-services.net/api/message-template",
        json=response_payload,
    )

    with pytest.raises(SystemExit):
        cli.main()

    out, _ = capsys.readouterr()
    assert out == json.dumps(response_payload) + "\n"


def test_file(capsys, cli, filesystem, mocked_responses):
    sys.argv = [
        "ritdu-slacker",
        "file",
        "--workspace",
        "myworkspace",
        "--channel",
        "mychannel",
        "--text",
        "Oops!",
        "--file",
        "/tmp/errorlog.txt",
        "--thread-uuid",
        "30D06A8F-B6BE-43BD-BE75-8716C26C1102",
    ]

    filesystem.create_file("/tmp/errorlog.txt", contents="This is a test")

    response_payload = {
        "ok": True,
        "channel": "C123456",
        "message": {
            "text": "Oops!",
            "type": "message",
        },
    }

    mocked_responses.add(
        responses.POST,
        "https://slacker.cube-services.net/api/message-template",
        json=response_payload,
    )

    with pytest.raises(SystemExit):
        cli.main()

    out, _ = capsys.readouterr()
    assert out == json.dumps(response_payload) + "\n"


ritdu_slacker/api.py
import json
from signal import signal, SIGINT
from requests import Session
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry


class SlackClient:
    def __init__(self, retries=5, backoff_factor=0.5, status_forcelist=(500, 502, 504)):
        self.retries = retries
        self.backoff_factor = backoff_factor
        self.status_forcelist = status_forcelist
        self.session = Session()
        self.retry = Retry(
            total=self.retries,
            read=self.retries,
            connect=self.retries,
            backoff_factor=self.backoff_factor,
            status_forcelist=self.status_forcelist,
        )
        self.adapter = HTTPAdapter(max_retries=self.retry)
        self.session.mount("http://", self.adapter)
        self.session.mount("https://", self.adapter)

    def message(
        self,
        text:str,
        workspace:str,
        channel:str,
        thread_uuid:str = '',
        message_uuid:str='',
        message_or_thread_uuid:str='',
        thread_broadcast:bool=False
    ):
        url = "https://slacker.cube-services.net/api/message-template"
        data = {
            "command": "SimpleMessage",
            "workspace": workspace,
            "channel": channel,
            "message_uuid": message_uuid,
            "thread_uuid": thread_uuid,
            "message_or_thread_uuid": message_or_thread_uuid,
            "thread_broadcast": thread_broadcast,
            "message": text,
            "fallback_message": text,
        }
        response = self.session.post(
            url=url, json=data, timeout=(10, 10)
        ).json()
        return response

    def file(
        self,
        file_bytes:str,
        text:str,
        workspace:str,
        channel:str,
        thread_uuid:str='',
        message_uuid:str='',
        message_or_thread_uuid:str='',
        thread_broadcast:bool=False,
    ):
        url = "https://slacker.cube-services.net/api/message-template"
        files = {"file": file_bytes}
        data = {
            "command": "SimpleMessage",
            "workspace": workspace,
            "channel": channel,
            "message_uuid": message_uuid,
            "thread_uuid": thread_uuid,
            "message_or_thread_uuid": message_or_thread_uuid,
            "thread_broadcast": thread_broadcast,
            "message": text,
            "fallback_message": text,
        }
        response = self.session.post(
            url=url, files=files, json=data, timeout=(10, 10)
        ).json()
        return response


ritdu_slacker/commands/commands.py
import base64
import logging, click, click_log, json
import requests, json, click, logging, os
from requests import Session
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.retry import Retry

logger = logging.getLogger()
logger.setLevel(logging.ERROR)
CONTEXT_SETTINGS = dict(help_option_names=["-h", "--help"])

status_map = {"ok": "#00FF00", "warning": "warning", "error": "danger"}


def requests_retry_session(
    retries=5,
    backoff_factor=0.5,
    status_forcelist=(500, 502, 504),
    session=None,
) -> Session:
    session = session or requests.Session()
    retry = Retry(
        total=retries,
        read=retries,
        connect=retries,
        backoff_factor=backoff_factor,
        status_forcelist=status_forcelist,
    )
    adapter = HTTPAdapter(max_retries=retry)
    session.mount("http://", adapter)
    session.mount("https://", adapter)
    return session


def post_message_to_slack(
    text, thread_uuid, message_uuid, workspace, channel, thread_broadcast=False
) -> dict:
    return (
        requests_retry_session()
        .post(
            url="https://slacker.cube-services.net/api/message-template",
            headers={"Content-Type": "application/json"},
            json={
                "command": "SimpleMessage",
                "workspace": workspace,
                "channel": channel,
                "message_uuid": message_uuid,
                "message_or_thread_uuid": "",
                "thread_uuid": thread_uuid,
                "thread_broadcast": thread_broadcast,
                "message": text,
                "fallback_message": text,
            },
            timeout=(10, 10),
        )
        .json()
    )


def post_file_to_slack(
    text,
    thread_uuid,
    message_uuid,
    workspace,
    channel,
    file_name,
    file_bytes,
    thread_broadcast=False,
) -> dict:
    return (
        requests_retry_session()
        .post(
            url="https://slacker.cube-services.net/api/message-template",
            headers={
                "Accept": "application/json",
            },
            files={"file": file_bytes},
            data={
                "json": json.dumps(
                    {
                        "command": "SimpleMessage",
                        "workspace": workspace,
                        "channel": channel,
                        "message_uuid": message_uuid,
                        "message_or_thread_uuid": "",
                        "thread_uuid": thread_uuid,
                        "thread_broadcast": thread_broadcast,
                        "message": text,
                        "fallback_message": text,
                    }
                )
            },
            timeout=(10, 10),
        )
        .json()
    )


def check_result(result):
    return result.get("ok")


@click.group(help="CLI tool send/update slack messages and send files")
@click.help_option("--help", "-h")
def main():
    pass


@click.command(
    help="Command to send message, reply to thread or reply and broadcast to thread"
)
@click.option("--debug", is_flag=True, help="Switch to debug logging")
@click.option("--text", "-m", default=None, required=True, help="text to send")
@click.option(
    "--workspace", "-w", default=None, required=True, help="slack workspace name"
)
@click.option("--channel", "-c", default=None, required=True, help="slack channel name")
@click.option(
    "--message-uuid",
    "-u",
    default=None,
    required=False,
    help="uuid of message to update",
)
@click.option(
    "--thread-uuid",
    "-t",
    default=None,
    required=False,
    help="thread uuid to send message to",
)
@click.option(
    "--thread-broadcast",
    "-b",
    is_flag=True,
    help="flag to broadcast message to thread",
)
def message(
    debug, text, thread_uuid, message_uuid, workspace, channel, thread_broadcast
):
    if debug:
        logger.setLevel(logging.DEBUG)
    logger.debug(f"Send text: {text}")
    result = post_message_to_slack(
        text=f"{text}",
        thread_uuid=thread_uuid if thread_uuid else "",
        message_uuid=message_uuid if message_uuid else "",
        workspace=workspace,
        channel=channel,
        thread_broadcast=thread_broadcast,
    )
    print(json.dumps(result))


@click.command(help="Command to send file to thread")
@click.option("--debug", is_flag=True, help="Switch to debug logging")
@click.option("--text", "-m", default=None, required=False, help="text to send")
@click.option(
    "--workspace", "-w", default=None, required=True, help="slack workspace name"
)
@click.option("--channel", "-c", default=None, required=True, help="slack channel name")
@click.option(
    "--message-uuid",
    "-u",
    default=None,
    required=False,
    help="uuid of message to update",
)
@click.option(
    "--thread-uuid",
    "-t",
    default=None,
    required=False,
    help="thread uuid to send message to",
)
@click.option(
    "--thread-broadcast",
    "-b",
    is_flag=True,
    help="flag to broadcast message to thread",
)
@click.option(
    "--file",
    "-f",
    default=None,
    required=True,
    help="file to send to slack, can be path to file or url",
)
def file(
    debug, text, thread_uuid, message_uuid, workspace, channel, file, thread_broadcast
):
    if debug:
        logger.setLevel(logging.DEBUG)
    logger.debug(f"Send file: {file}")

    file_basename = os.path.basename(file)
    with open(file, "rb") as f:
        result = post_file_to_slack(
            text=f"{text}" if text else f"File: {file_basename}",
            thread_uuid=thread_uuid if thread_uuid else "",
            message_uuid=message_uuid if message_uuid else "",
            workspace=workspace,
            channel=channel,
            file_name=file_basename,
            file_bytes=f,
            thread_broadcast=thread_broadcast,
        )
    print(json.dumps(result))
